analyzers/src/SonarAnalyzer.CSharp/Rules/LoopsAndLinq.cs

```diff
@@ -0,0 +1,137 @@
+
+        private static readonly DiagnosticDescriptor Rule = DiagnosticDescriptorBuilder.GetDescriptor(DiagnosticId, MessageFormat, RspecStrings.ResourceManager);
+
+        public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics => ImmutableArray.Create(Rule);
+
+        protected override void Initialize(SonarAnalysisContext context) =>
+            context.RegisterSyntaxNodeActionInNonGenerated(c =>
+                {
+                    var forEachStatementSyntax = (ForEachStatementSyntax)c.Node;
+                    if (CanBeSimplifiedUsingWhere(forEachStatementSyntax.Statement))
+                    {
+                        c.ReportDiagnosticWhenActive(Diagnostic.Create(Rule, forEachStatementSyntax.Expression.GetLocation(), WhereMessageFormat));
+                    }
+                    else
+                    {
+                        CheckIfCanBeSimplifiedUsingSelect(forEachStatementSyntax, c);
+                    }
+                },
+                SyntaxKind.ForEachStatement);
+
+        private static bool CanBeSimplifiedUsingWhere(SyntaxNode statement) =>
+            GetIfStatement(statement) is { } ifStatementSyntax
+            && CanIfStatementBeMoved(ifStatementSyntax);
+
+        private static IfStatementSyntax GetIfStatement(SyntaxNode node) =>
+            node switch
+            {
+                IfStatementSyntax ifStatementSyntax => ifStatementSyntax,
+                BlockSyntax blockSyntax when blockSyntax.ChildNodes().Count() == 1 => GetIfStatement(blockSyntax.ChildNodes().Single()),
+                _ => null
+            };
+
+        private static bool CanIfStatementBeMoved(IfStatementSyntax ifStatementSyntax) =>
+            ifStatementSyntax.Else == null
+            && ifStatementSyntax.Condition is InvocationExpressionSyntax invocationExpressionSyntax
+            && !invocationExpressionSyntax.DescendantNodes()
+                                          .OfType<ArgumentSyntax>()
+                                          .Any(argument => argument.RefOrOutKeyword.IsAnyKind(SyntaxKind.OutKeyword, SyntaxKind.RefKeyword));
+
+        private static void CheckIfCanBeSimplifiedUsingSelect(ForEachStatementSyntax forEachStatementSyntax, SyntaxNodeAnalysisContext c)
+        {
+            // There are multiple scenarios where the code can be simplified using LINQ.
+            // For simplicity, in the first version of the rule, we consider that Select() can be used
+            // only when a single property from the foreach variable is used.
+            // The issue is raised if:
+            //  - the property is used more than once
+            //  - the property is the right side of a variable declaration
+            var declaredSymbol = c.SemanticModel.GetDeclaredSymbol(forEachStatementSyntax);
+            var accessedProperties = new Dictionary<ISymbol, UsageStats>();
+
+            foreach (var identifierSyntax in GetIdentifiers(forEachStatementSyntax))
```

Comment:
```suggestion
            foreach (var identifierInBody in GetIdentifiersInBody(forEachStatementSyntax))
```

- `identifierSyntax` doesn't convey a lot of meaning, it's quite generic

- `GetIdentifiers` is looking in the `Statement` which is actually the body of the foreach (BlockStatement or ExpressionStatement if no block) - to make it clear it's not looking into the `foreach (...)` part of the syntax